components/registry/org.wso2.carbon.registry.extensions/src/main/java/org/wso2/carbon/registry/extensions/handlers/XSDMediaTypeHandler.java

```diff
@@ -138,28 +139,51 @@ public void put(RequestContext requestContext) throws RegistryException {
                 }
                 oldResourcePath = resourcePath; // keep the old resource path.
             }
-
             WSDLValidationInfo validationInfo = null;
-            Object resourceContent = resource.getContent();
-            if (resourceContent instanceof String) {
-                resourceContent = RegistryUtils.encodeString(((String) resourceContent));
-                resource.setContent(resourceContent);
-            }
-            if (resourceContent instanceof byte[]) {
+            String savedName;
+
+            requestContext.setSourceURL(
+                    requestContext.getResource().getProperty(CommonConstants.SOURCEURL_PARAMETER_NAME));
+            String sourceURL = requestContext.getSourceURL();
+
+            if (StringUtils.isNotBlank(sourceURL)) {
+                if (requestContext.getSourceURL().toLowerCase()
+                        .startsWith("file:")) {
+                    String msg = "The source URL must not be file in the server's local file system";
+                    throw new RegistryException(msg);
+                }
                 try {
-                    InputStream in = new ByteArrayInputStream((byte[]) resourceContent);
                     if (!disableSchemaValidation) {
-                        validationInfo =
-                                SchemaValidator.validate(new XMLInputSource(null, null, null, in,
-                                        null));
+                        validationInfo = SchemaValidator.validate(new XMLInputSource(null, sourceURL, null));
                     }
                 } catch (Exception e) {
-                    throw new RegistryException("Exception occured while validating the schema", e);
+                    // Since SchemaValidator.validate method is throwing Exception need to catch it here
+                    throw new RegistryException("Exception occurred while validating the schema", e);
                 }
-            }
 
-            String savedName = processSchemaUpload(requestContext, resourcePath, validationInfo);
+                savedName = processSchemaImport(requestContext, resourcePath, validationInfo);
 
+            } else {
+                Object resourceContent = resource.getContent();
+                if (resourceContent instanceof String) {
+                    resourceContent = RegistryUtils.encodeString(((String) resourceContent));
+                    resource.setContent(resourceContent);
+                }
+                if (resourceContent instanceof byte[]) {
+                    try {
+                        InputStream in = new ByteArrayInputStream((byte[]) resourceContent);
+                        if (!disableSchemaValidation) {
+                            validationInfo = SchemaValidator.
+                                    validate(new XMLInputSource(null, null, null, in, null));
```

Comment:
It would be better if we can add inline comments to make the code more readable. In this line it is not clear why four null parameters are parsed.
Thanks